Ihc command specific outbinding

smerschjohann opened the pull request:
This modification allows to specify different resources and the wanted value related to the called command.

# Configuration Syntax
## Before
```
ihc=">123456"
```

Only a single resourceId could be referenced. Values were automatically determined based on the item state.

## After

```
ihc=">[UP:12345:100],>[DOWN:0x9bcd:100],>[STOP:99999]"
```

Syntax:  >[COMMAND:RESOURCE_ID(:VALUE)]

1. Command which is called
2. The resource id to use (hex or int)
3. optional
    * if not specified the value will be determined like before
    * 1: OnOffType.ON is sent to the resource
    * 0: OnOffType.OFF is sent to the resource
    * value > 1: Sequence is sent: OnOfType.ON, sleep(value), OnOffType.OFF

buildhive commented:
[openhab » openhab #2638](https://buildhive.cloudbees.com/job/openhab/job/openhab/2638/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

teichsta commented on bundles/binding/org.openhab.binding.ihc/src/main/java/org/openhab/binding/ihc/internal/IhcGenericBindingProvider.java:
```diff
@@ -87,50 +94,92 @@ public void processBindingConfiguration(String context, Item item,
+					String resourceCommand = split.substring(1);
+					
+					final String commandRegex = "\\[([\\w*]+):(0x[0-9a-fA-F]+|\\d+)(?::(\\d+)){0,1}\\]"; 
+					Pattern p = Pattern.compile(commandRegex);
```
please extract the Regex String into a precompiled final static Pattern

teichsta commented:
@paulianttila could you please have a look at these changes? Thanks, Thomas E.-E.

paulianttila commented on bundles/binding/org.openhab.binding.ihc/src/main/java/org/openhab/binding/ihc/internal/IhcBinding.java:
```diff
@@ -396,28 +399,79 @@ private void updateResource(String itemName, Type type, boolean updateOnlyOutBin
+							logger.debug("Item '{}' trigger started",
+									itemName);
+							
+							Thread.sleep(val);
```
This sleep duration probably should be limited? Long sleep will block another resource id updates.

paulianttila commented on bundles/binding/org.openhab.binding.ihc/src/main/java/org/openhab/binding/ihc/internal/IhcBinding.java:
```diff
@@ -396,28 +399,79 @@ private void updateResource(String itemName, Type type, boolean updateOnlyOutBin
+								.getDefinitionTypeID());
+					}
+	
+					Integer val = provider.getValue(itemName, (Command)type);
```
Could you add comment here, how this val and trigger special case works. Without comment, code could be little bit confusing  afterwards.

paulianttila commented on bundles/binding/org.openhab.binding.ihc/src/main/java/org/openhab/binding/ihc/internal/IhcGenericBindingProvider.java:
```diff
@@ -87,50 +94,92 @@ public void processBindingConfiguration(String context, Item item,
+					if (configParts.length == 2)
+						config.refreshInterval = Integer.parseInt(configParts[1]);
+				} else {
+					throw new BindingConfigParseException("Only out binding or normal binding supported at this time");
```
Is this exception message correct?

paulianttila commented on bundles/binding/org.openhab.binding.ihc/src/main/java/org/openhab/binding/ihc/internal/IhcGenericBindingProvider.java:
```diff
@@ -87,50 +94,92 @@ public void processBindingConfiguration(String context, Item item,
 			} else {
-				config.resourceId = Integer.parseInt(resourceId);
+				if(splittedCommands.length < 2) {
+					String[] configParts = bindingConfig.trim().split(":");
```
Length of the configParts should be checked. If length is bigger than 2, exception should be thrown. In old code this was done in line 93.

smerschjohann commented:
okay I've changed the code at the given positions. In addition I've added "In-Bindings".

You can now combine In- and Out-Bindings together. Example ihc="<12345,>[*:54321]". The corresponding item will get its updates from the 12345 resourceId, and update 54321 by any given command.

buildhive commented:
[openhab » openhab #2707](https://buildhive.cloudbees.com/job/openhab/job/openhab/2707/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

paulianttila commented on bundles/binding/org.openhab.binding.ihc/src/main/java/org/openhab/binding/ihc/internal/IhcGenericBindingProvider.java:
```diff
@@ -169,10 +276,25 @@ public int getRefreshInterval(String itemName) {
+	public boolean hasInBinding(String itemName) {
+		IhcBindingConfig config = (IhcBindingConfig) bindingConfigs.get(itemName);
+		
+		return config != null ? config.resourceId > 0 : null;
```
Does this really work? Resource id is always bigger than 0, so how old "out binding only" situation is now detected?

smerschjohann replied:
I changed it in that way, that the "In-Binding" will use the resourceId from IhcBindingConfig. Every Out-Going Binding ("1234", ">1234") will use the IhcOutCommandConfig and will get an entry in the outCommandMap.
An out-only binding does not set the resourceId in IhcBindingConfig and therefore it is assigned with zero and hasInBinding would return false.

paulianttila replied:
Maybe I didn't fully understand your explanation, but for old style out binding, resource id is always set (see IhcGenericBindingProvider.java line 118).

smerschjohann replied:
There exists two classes: IhcOutCommandConfig (Configuration for an out configuration, this is used to configure resources for outgoing "commands").

In addition there is the IhcBindingConfig class. This class groups all configurations available to an item. The resourceId in this class is used for the InBinding and is by default 0.

When you look at the line 118 you will see that the resourceId is set at the IhcOutCommandConfig and not on the IhcBindingConfig and therefore the resourceId responsible for the In-Binding remains zero.

paulianttila replied:
OK, now I understand :) It's sometimes difficult to follow code just in PR diff, sorry about that.

paulianttila commented on bundles/binding/org.openhab.binding.ihc/src/main/java/org/openhab/binding/ihc/internal/IhcBinding.java:
```diff
@@ -552,14 +608,14 @@ public void resourceValueUpdateReceived(EventObject event,
 				if (value.getResourceID() == resourceId) {
 
-					if (provider.isOutBindingOnly(itemName)) {
+					if (!provider.hasInBinding(itemName)) {
```
See my comment in hasInBinding method implementation

paulianttila commented:
Btw, you should use eclipse code formatter to format you code changes according to openhab code styling.

smerschjohann commented:
I reformatted the code with the code style. Don't know if I really like that style (not really readable when the function call is in the next line and not indented wide enough), but well it is existing code and I agree it should be one code style in a file.

buildhive commented:
[openhab » openhab #2720](https://buildhive.cloudbees.com/job/openhab/job/openhab/2720/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

teichsta commented:
looks good to me altogether now. If you could now squash your commits into one it will be a pleasure to merge into master :-) Thanks, Thomas E.-E.

buildhive commented:
[openhab » openhab #2730](https://buildhive.cloudbees.com/job/openhab/job/openhab/2730/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)
